### ISSUE TITLE
[Suggestion] SMM -> SMMDB API routing

### ISSUE THREAD

**Issue by RedDuckss** [decaf-emu/decaf-emu#453]
I suggested this to several people in the Cemu community, due to the advent of CemuHook which allows developers to run 3rd party code in Cemu, but everyone has declined so far. Figured the suggestion would best be suggested here since Decaf is open source.

There is a very popular Super Mario Maker fan site called SMMDB. It started off as a SMM level sharing site for Cemu specifically but has grown to supper Super Mario 64 Maker, Cemu and WiiU/3DS Homebrew.

The site also offers a rich API for searching, downloading, start/rating and uploading courses.

Would it be possible to have some specific checks in Decaf that when running Super Mario Maker and the user tries to use the online services, those requests are routed as API requests to SMMDB, to add online SMM functionality to the emulator?

**Comment by CuriousTommy** [comment]
> Would it be possible to have some specific checks in Decaf that when running Super Mario Maker and the user tries to use the online services, those requests are routed as API requests to SMMDB, to add online SMM functionality to the emulator?

Honestly, it would be a better idea to modify the game's executable to access the SMMDB. That way, it can also work on a real Wii U too. I personally hate seeing emulator specific mods.

**Comment by VelocityRa** [comment]
Does the SMMDB API reflect Nintendo's actual Mario Maker API? Because if so then @CuriousTommy's solution would be much better. If not, then a translation layer will be needed (doesn't have to be an game/emulator-specific mod).

**Comment by RedDuckss** [comment]
> Honestly, it would be a better idea to modify the game's executable to access the SMMDB. That way, it can also work on a real Wii U too. I personally hate seeing emulator specific mods.

I made my suggestion in the SMMDB Discord server a while back and a few days after someone was talking about doing this exact thing (modifying the rom itself). He said:
```
I might be able to make mario maker read courses on SMMDB
i found what the .rpl file is used for
in the code folder

the Cemu-SMMDB would be assesible ingame

where you can play levels like it was from the game on the wii u
and this would in theory work with cemu

we would need to clone these web addresses and clone their behavior: 
https://wup-ama.app.nintendo.net/api/v1/pickup/easy
 
https://wup-ama.app.nintendo.net/api/v1/pickup/super_expert
 
https://wup-ama.app.nintendo.net/api/v1/pickup/normal
 
https://wup-ama.app.nintendo.net/api/v1/pickup/expert
 and 
https://wup-ama.app.nintendo.net/api/v1/playlist
```

So I guess that's a step in the right direction.

> Does the SMMDB API reflect Nintendo's actual Mario Maker API?

No not that I'm aware of. Though you would probably want to ask @Tarnadas since he's the one behind SMMDB

**Comment by Tarnadas** [comment]
No it does not reflect Nintendo's API, so we would definitely need such a translation layer or if someone can tell me how these API endpoints work, I can try to copy them

**Comment by VelocityRa** [comment]
Yeah we'd need to reverse engineer them. A Wii U and a way to capture the requests would def. make things easier, but it would still be pretty time consuming.
Maybe someone could share a packet dump of a play session? Would make it possible for folks that don't have a Wii U/SMM to try and figure out a few things about the API.

**Comment by GRAnimated** [comment]
https://www.reddit.com/r/WiiUHacks/comments/6yzdy9/i_started_a_mod_for_smm_which_loads_levels_from/ I made a page on /r/WiiUHacks about it a couple days ago, seems like I have found all of the behavior, but I don't have the way they store their courses and images on their server. (The wup.ama servers are all redirects I found to an API server and a cloudfront server, but it seems as if the cloudfront server only stores images.)

**Comment by VelocityRa** [comment]
>seems like I have found all of the behavior

So you have the client certs and the headers too? Can you successfully fake a connection to the endpoints (or at least get around the 400 error)?

>but I don't have the way they store their courses and images on their server

You mean the file format for them? Presumably you don't _need_ to know their format, if we can get a hold of the data, the game will know what to do with it.

**Comment by GRAnimated** [comment]
I've already got past the 400 errors, I just need to know the way they store them, so like which directory. I also already know the file formats.

**Comment by RedDuckss** [comment]
I know this topic is really old now (I forgot about it until just now), however since it isn't closed yet I will assume there is still some interest here.

We have made great progress on breaking down a lot of areas of the SMM server schema, and I have begun replicating it and making a custom server. I have documented everything here in this GitHub repo https://github.com/RedDuckss/csms

@GRAnimated was able to mod an RPX rom of SMM to connect to a custom server of mine, however the tunnel was refused and didn't work. We think this has to do with the SSL certs used on my server (I use standard Let's Encrypt SSL certs) but we aren't 100% sure if that's the cause or if it's something totally different.

**Comment by VelocityRa** [comment]
@RedDuckss @GRAnimated 
Nice work guys.

The issue you mentioned sounds a lot like an issue I've been facing with a very similar project. Turned out that it was because of [certificate pinning](https://security.stackexchange.com/questions/29988/what-is-certificate-pinning). Still haven't solved it yet, but at least I know why it happens...

So if this is indeed the issue, your only solution might be patching the executable to get around this extra layer of validation.
This is all assuming you have actually [trusted the Let's Encrypt certificates you're using](https://gbatemp.net/threads/tutorial-add-custom-root-certs-to-the-wii-us-browser.468201) on the device itself of course. If they're not root certs (or signed by root certs) no way the device will trust them (unless you patch the fw or something).

Btw you're on the Nintendo Homebrew Discord right? We can talk there if you want.

**Comment by RedDuckss** [comment]
I'm not in the Homebrew Discord. I believe I was at a time but left, mind shooting me a link again?

**Comment by VelocityRa** [comment]
Sure: https://discord.gg/C29hYvh

**Comment by mrjvs** [comment]
@VelocityRa What is your discord username? (If you are wondering who I am, I am also working with redducks and granimated on the custom super mario maker server)

**Comment by VelocityRa** [comment]
It's Vel0cityX#6394.

**Comment by exjam** [comment]
Awesome - i'll leave this open for now though as it seems it would require decaf to have support for being pointed at the SMMDB server.

I can see you are hooking cemu to force that for now, I am sure we can do better in an open source emulator once the time comes 😛 

Although I still have to fix the fact that Super Mario Maker is broken first :D, I think there is a bug which corrupts the PM4 stream and it seems to affect a few games including this one unfortunately. I will probably rewrite all the PM4 serialisation because its a mess anyway.

**Comment by mrexodia** [comment]
Yeah the hook is to redirect the DNS queries. The same could be accomplished with editing the hosts file, but it’s less user friendly 🙂
